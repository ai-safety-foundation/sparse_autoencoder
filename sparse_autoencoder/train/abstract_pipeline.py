"""Abstract pipeline."""
from abc import ABC, abstractmethod
from collections.abc import Iterable
from pathlib import Path
from typing import final

import torch
from torch.utils.data import DataLoader
from tqdm.auto import tqdm
from transformer_lens import HookedTransformer
import wandb

from sparse_autoencoder.activation_resampler.abstract_activation_resampler import (
    AbstractActivationResampler,
)
from sparse_autoencoder.activation_store.tensor_store import TensorActivationStore
from sparse_autoencoder.autoencoder.model import SparseAutoencoder
from sparse_autoencoder.loss.abstract_loss import AbstractLoss
from sparse_autoencoder.metrics.metrics_container import MetricsContainer, default_metrics
from sparse_autoencoder.metrics.resample.abstract_resample_metric import ResampleMetricData
from sparse_autoencoder.optimizer.abstract_optimizer import AbstractOptimizerWithReset
from sparse_autoencoder.source_data.abstract_dataset import SourceDataset, TorchTokenizedPrompts
from sparse_autoencoder.tensor_types import (
    NeuronActivity,
)


class AbstractPipeline(ABC):
    """Pipeline for training a Sparse Autoencoder on TransformerLens activations.

    Includes all the key functionality to train a sparse autoencoder, with a specific set of
        hyperparameters.
    """

    activation_resampler: AbstractActivationResampler | None
    """Activation resampler to use."""

    autoencoder: SparseAutoencoder
    """Sparse autoencoder to train."""

    cache_name: str
    """Name of the cache to use in the source model (hook point)."""

    layer: int
    """Layer to get activations from with the source model."""

    log_frequency: int
    """Frequency at which to log metrics (in steps)."""

    loss: AbstractLoss
    """Loss function to use."""

    metrics: MetricsContainer
    """Metrics to use."""

    optimizer: AbstractOptimizerWithReset
    """Optimizer to use."""

    progress_bar: tqdm | None
    """Progress bar for the pipeline."""

    source_data: Iterable[TorchTokenizedPrompts]
    """Iterable over the source data."""

    source_dataset: SourceDataset
    """Source dataset to generate activation data from (tokenized prompts)."""

    source_model: HookedTransformer
    """Source model to get activations from."""

    total_training_steps: int = 1
    """Total number of training steps state."""

    @final
    def __init__(  # noqa: PLR0913
        self,
        activation_resampler: AbstractActivationResampler | None,
        autoencoder: SparseAutoencoder,
        cache_name: str,
        layer: int,
        loss: AbstractLoss,
        optimizer: AbstractOptimizerWithReset,
        source_dataset: SourceDataset,
        source_model: HookedTransformer,
        checkpoint_directory: Path | None = None,
        log_frequency: int = 100,
        metrics: MetricsContainer = default_metrics,
        source_data_batch_size: int = 12,
    ):
        """Initialize the pipeline.

        Args:
            activation_resampler: Activation resampler to use.
            autoencoder: Sparse autoencoder to train.
            cache_name: Name of the cache to use in the source model (hook point).
            layer: Layer to get activations from with the source model.
            loss: Loss function to use.
            optimizer: Optimizer to use.
            source_dataset: Source dataset to get data from.
            source_model: Source model to get activations from.
            checkpoint_directory: Directory to save checkpoints to.
            log_frequency: Frequency at which to log metrics (in steps)
            metrics: Metrics to use.
            source_data_batch_size: Batch size for the source data.
        """
        self.activation_resampler = activation_resampler
        self.autoencoder = autoencoder
        self.cache_name = cache_name
        self.checkpoint_directory = checkpoint_directory
        self.layer = layer
        self.log_frequency = log_frequency
        self.loss = loss
        self.metrics = metrics
        self.optimizer = optimizer
        self.source_data_batch_size = source_data_batch_size
        self.source_dataset = source_dataset
        self.source_model = source_model

        source_dataloader = source_dataset.get_dataloader(source_data_batch_size)
        self.source_data = self.stateful_dataloader_iterable(source_dataloader)

    @abstractmethod
    def generate_activations(self, store_size: int) -> TensorActivationStore:
        """Generate activations.

        Args:
            store_size: Number of activations to generate.

        Returns:
            Activation store for the train section.
        """

    @abstractmethod
    def train_autoencoder(
        self, activation_store: TensorActivationStore, train_batch_size: int
    ) -> NeuronActivity:
        """Train the sparse autoencoder.

        Args:
            activation_store: Activation store from the generate section.
            train_batch_size: Train batch size.

        Returns:
            Number of times each neuron fired.
        """

    @final
    def resample_neurons(
        self,
        activation_store: TensorActivationStore,
        neuron_activity_sample_size: int,
        neuron_activity: NeuronActivity,
        train_batch_size: int,
    ) -> None:
        """Resample dead neurons.

        Args:
            activation_store: Activation store.
            neuron_activity_sample_size: Sample size for resampling.
            neuron_activity: Number of times each neuron fired.
            train_batch_size: Train batch size (also used for resampling).
        """
        if self.activation_resampler is not None:
            # Get the updates
            parameter_updates = self.activation_resampler.resample_dead_neurons(
                activation_store=activation_store,
                autoencoder=self.autoencoder,
                loss_fn=self.loss,
                neuron_activity=neuron_activity,
                train_batch_size=train_batch_size,
                neuron_activity_sample_size=neuron_activity_sample_size,
            )

            # Update the weights and biases
            self.autoencoder.encoder.update_dictionary_vectors(
                parameter_updates.dead_neuron_indices,
                parameter_updates.dead_encoder_weight_updates,
            )
            self.autoencoder.encoder.update_bias(
                parameter_updates.dead_neuron_indices,
                parameter_updates.dead_encoder_bias_updates,
            )
            self.autoencoder.decoder.update_dictionary_vectors(
                parameter_updates.dead_neuron_indices,
                parameter_updates.dead_decoder_weight_updates,
            )

            # Log any metrics
            with torch.no_grad():
                metrics = {}
                if wandb.run is not None:
                    for metric in self.metrics.resample_metrics:
                        calculated = metric.calculate(
                            ResampleMetricData(
                                neuron_activity=neuron_activity,
                            )
                        )
                        metrics.update(calculated)
                    wandb.log(metrics, commit=False)

            # Reset the optimizer (TODO: Consider resetting just the relevant parameters)
            self.optimizer.reset_state_all_parameters()

    @abstractmethod
    def validate_sae(self) -> None:
        """Get validation metrics."""

    @final
    def save_checkpoint(self) -> None:
        """Save the model as a checkpoint."""
        if self.checkpoint_directory:
            file_path: Path = (
                self.checkpoint_directory / f"sae_state_dict-{self.total_training_steps}.pt"
            )
            torch.save(self.autoencoder.state_dict(), file_path)

    @final
    def run_pipeline(
        self,
        train_batch_size: int,
        max_store_size: int,
        max_activations: int,
        resample_frequency: int,
        validate_frequency: int | None = None,
        checkpoint_frequency: int | None = None,
    ) -> None:
        """Run the full training pipeline.

        Args:
            train_batch_size: Train batch size.
            max_store_size: Maximum size of the activation store.
            max_activations: Maximum total number of activations to train on (the original paper
                used 8bn, although others have had success with 100m+).
            resample_frequency: Frequency at which to resample dead neurons (the original paper used
                every 200m).
            validate_frequency: Frequency at which to get validation metrics.
            checkpoint_frequency: Frequency at which to save a checkpoint.
        """
        last_resampled: int = 0
        neuron_activity_sample_size: int = 0
        last_validated: int = 0
        last_checkpoint: int = 0
        total_activations: int = 0
        neuron_activity: NeuronActivity = torch.zeros(self.autoencoder.n_learned_features)

        # Get the store size
        store_size: int = max_store_size - max_store_size % (
            self.source_data_batch_size * self.source_dataset.context_size
        )

        with tqdm(
            desc="Activations trained on",
            total=max_activations,
        ) as progress_bar:
            for _ in range(0, max_activations, store_size):
                # Generate
                progress_bar.set_postfix({"stage": "generate"})
                activation_store: TensorActivationStore = self.generate_activations(store_size)

                # Update the counters
                num_activation_vectors_in_store = len(activation_store)
                last_resampled += num_activation_vectors_in_store
                last_validated += num_activation_vectors_in_store
                last_checkpoint += num_activation_vectors_in_store
                total_activations += num_activation_vectors_in_store
                if wandb.run is not None:
                    wandb.log({"total_activations": total_activations}, commit=False)

                # Train
                progress_bar.set_postfix({"stage": "train"})
                batch_neuron_activity: NeuronActivity = self.train_autoencoder(
                    activation_store, train_batch_size=train_batch_size
                )
                detached_neuron_activity = batch_neuron_activity.detach().cpu()
                is_second_half_resample: bool = last_resampled > resample_frequency / 2
                if is_second_half_resample:
                    neuron_activity_sample_size += num_activation_vectors_in_store
                    neuron_activity.add_(detached_neuron_activity)

                # Resample dead neurons (if needed)
                progress_bar.set_postfix({"stage": "resample"})
                if last_resampled >= resample_frequency and self.activation_resampler is not None:
                    self.resample_neurons(
                        activation_store=activation_store,
                        neuron_activity_sample_size=neuron_activity_sample_size,
                        neuron_activity=neuron_activity,
                        train_batch_size=train_batch_size,
                    )

                    # Reset
                    last_resampled = 0
                    neuron_activity_sample_size = 0
                    neuron_activity.zero_()

                # Get validation metrics (if needed)
                progress_bar.set_postfix({"stage": "validate"})
                if validate_frequency is not None and last_validated > validate_frequency:
                    self.validate_sae()
                    self.last_validated = 0

                # Checkpoint (if needed)
                progress_bar.set_postfix({"stage": "checkpoint"})
                if checkpoint_frequency is not None and last_checkpoint > checkpoint_frequency:
                    self.last_checkpoint = 0
                    self.save_checkpoint()

                # Update the progress bar
                progress_bar.update(store_size)

    @staticmethod
    def stateful_dataloader_iterable(
        dataloader: DataLoader[TorchTokenizedPrompts]
    ) -> Iterable[TorchTokenizedPrompts]:
        """Create a stateful dataloader iterable.

        Create an iterable that maintains it's position in the dataloader between loops.

        Examples:
            Without this, when iterating over a DataLoader with 2 loops, each loop get the same data
            (assuming shuffle is turned off). That is to say, the second loop won't maintain the
            position from where the first loop left off.

            >>> from datasets import Dataset
            >>> from torch.utils.data import DataLoader
            >>> def gen():
            ...     yield {"int": 0}
            ...     yield {"int": 1}
            >>> data = DataLoader(Dataset.from_generator(gen))
            >>> next(iter(data))["int"], next(iter(data))["int"]
            (tensor([0]), tensor([0]))

            By contrast if you create a stateful iterable from the dataloader, each loop will get
            different data.

            >>> iterator = AbstractPipeline.stateful_dataloader_iterable(data)
            >>> next(iterator)["int"], next(iterator)["int"]
            (tensor([0]), tensor([1]))

        Args:
            dataloader: PyTorch DataLoader.

        Returns:
            Stateful iterable over the data in the dataloader.

        Yields:
            Data from the dataloader.
        """
        yield from dataloader
