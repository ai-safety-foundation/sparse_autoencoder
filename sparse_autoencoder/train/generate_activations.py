"""Generate activations for training a Sparse Autoencoder."""
from collections.abc import Iterable
from functools import partial

from jaxtyping import Int
import torch
from torch import Tensor
from tqdm.auto import tqdm
from transformer_lens import HookedTransformer

from sparse_autoencoder.activation_store.base_store import (
    ActivationStore,
    StoreFullError,
)
from sparse_autoencoder.source_data.abstract_dataset import TorchTokenizedPrompts
from sparse_autoencoder.src_model.store_activations_hook import store_activations_hook


def generate_activations(
    model: HookedTransformer,
    layer: int,
    cache_name: str,
    store: ActivationStore,
    source_data: Iterable[TorchTokenizedPrompts],
    context_size: int,
    batch_size: int,
    num_items: int,
    device: torch.device | None = None,
) -> None:
    """Generate activations for training a Sparse Autoencoder.

    Generates activations and updates the activation store in place.

    Warning:
    This function is a little confusing as it uses side effects. The way it works is to add a hook
    to the model, which will automatically store activations every time the model runs. When it has
    filled up the store to the desired size, it will return `None`. Your activations will then be
    ready in the `store` object that you passed to this function (i.e. it is updated in place). This
    approach is used as it depends on TransformerLens which uses side effects to get the
    activations.

    Args:
        model: The model to generate activations for.
        layer: The layer that you are hooking into to get activations. This is used to stop the
            model at this point rather than generating all remaining activations and logits.
        cache_name: The name of the cache hook point to get activations from. Examples include
            [`hook_embed` `hook_pos_embed`, `blocks.0.hook_resid_pre`, `blocks.0.ln1.hook_scale`,
            `blocks.0.ln1.hook_normalized`, `blocks.0.attn.hook_q`, `blocks.0.attn.hook_k`,
            `blocks.0.attn.hook_v`, `blocks.0.attn.hook_attn_scores`, `blocks.0.attn.hook_pattern`,
            `blocks.0.attn.hook_z`, `blocks.0.hook_attn_out`, `blocks.0.hook_resid_mid`,
            `blocks.0.ln2.hook_scale`, `blocks.0.ln2.hook_normalized`, `blocks.0.mlp.hook_pre`,
            `blocks.0.mlp.hook_post`, `blocks.0.hook_mlp_out`, `blocks.0.hook_resid_post`].
        store: The activation store to use.
        source_data: Stateful iterator that yields batches of data to generate activations.
        context_size: Number of tokens in each prompt.
        batch_size: Size of each batch.
        num_items: Number of activation vectors to generate. This is an approximate rather
            than strict limit.
        device: Device to run the model on.
    """
    if isinstance(device, torch.device):
        model.to(device, print_details=False)

    # Add the hook to the model (will automatically store the activations every time the model runs)
    model.remove_all_hook_fns()
    hook = partial(store_activations_hook, store=store)
    model.add_hook(cache_name, hook)

    # Get the input dimensions for logging
    activations_per_batch: int = context_size * batch_size
    total: int = num_items - num_items % activations_per_batch

    # Loop through the dataloader until the store reaches the desired size
    with torch.no_grad(), tqdm(
        desc="Generate Activations",
        total=total,
        colour="green",
        position=1,
        leave=False,
        dynamic_ncols=True,
    ) as progress_bar:
        for batch in source_data:
            try:
                input_ids: Int[Tensor, "batch pos"] = batch["input_ids"].to(device)
                model.forward(input_ids, stop_at_layer=layer + 1)  # type: ignore (TLens is typed incorrectly)
                progress_bar.update(activations_per_batch)

            # Break the loop if the store is full
            except StoreFullError:
                break

            if len(store) >= total:
                return

        progress_bar.close()
