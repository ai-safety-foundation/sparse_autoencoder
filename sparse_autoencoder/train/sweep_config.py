"""Sweep Config."""
from dataclasses import asdict, dataclass, field
from typing import Any

from sparse_autoencoder.train.utils.wandb_sweep_types import (
    Method,
    Metric,
    Parameter,
    Parameters,
    WandbSweepConfig,
)


# NOTE: This must be kept in sync with SweepParametersRuntime
@dataclass(frozen=True)
class SweepParameterConfig(Parameters):
    """Sweep Parameter Config."""

    lr: Parameter[float] | None
    """Adam Learning Rate."""

    adam_beta_1: Parameter[float] | None
    """Adam Beta 1.

    The exponential decay rate for the first moment estimates (mean) of the gradient.
    """

    adam_beta_2: Parameter[float] | None
    """Adam Beta 2.

    The exponential decay rate for the second moment estimates (variance) of the gradient.
    """

    adam_epsilon: Parameter[float] | None
    """Adam Epsilon.

    A small constant for numerical stability.
    """

    adam_weight_decay: Parameter[float] | None
    """Adam Weight Decay.

    Weight decay (L2 penalty).
    """

    l1_coefficient: Parameter[float] | None
    """L1 Penalty Coefficient.

    The L1 penalty is the absolute sum of learned (hidden) activations, multiplied by this constant.
    The penalty encourages sparsity in the learned activations. This loss penalty can be reduced by
    using more features, or using a lower L1 coefficient.

    Default values from the [original
    paper](https://transformer-circuits.pub/2023/monosemantic-features/index.html).
    """

    batch_size: Parameter[int] | None
    """Batch size.

    Used in SAE Forward Pass."""


# NOTE: This must be kept in sync with SweepParameterConfig
@dataclass(frozen=True)
class SweepParametersRuntime(dict[str, Any]):
    """Sweep parameter runtime values."""

    lr: float = 0.001

    adam_beta_1: float = 0.9

    adam_beta_2: float = 0.999

    adam_epsilon: float = 1e-8

    adam_weight_decay: float = 0.0

    l1_coefficient: float = 0.001

    batch_size: int = 4096

    def to_dict(self) -> dict[str, Any]:
        """Return dict representation of this object."""
        return asdict(self)


@dataclass(frozen=True)
class SweepConfig(WandbSweepConfig):
    """Sweep Config."""

    parameters: SweepParameterConfig

    method: Method = Method.grid

    metric: Metric = field(default_factory=lambda: Metric(name="loss"))

    def to_dict(self) -> dict[str, Any]:
        """Return dict representation of this object."""
        dict_representation = asdict(self)

        # Convert StrEnums to strings
        dict_representation["method"] = dict_representation["method"].value

        return dict_representation
