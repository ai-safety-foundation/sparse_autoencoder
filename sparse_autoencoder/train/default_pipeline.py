"""Default pipeline."""
from functools import partial
from typing import final

import torch
from torch.utils.data import DataLoader
import wandb

from sparse_autoencoder.activation_resampler.abstract_activation_resampler import (
    ParameterUpdateResults,
)
from sparse_autoencoder.activation_store.base_store import ActivationStore
from sparse_autoencoder.activation_store.tensor_store import TensorActivationStore
from sparse_autoencoder.src_model.store_activations_hook import store_activations_hook
from sparse_autoencoder.tensor_types import BatchTokenizedPrompts, NeuronActivity
from sparse_autoencoder.train.abstract_pipeline import AbstractPipeline
from sparse_autoencoder.train.resample_neurons import resample_dead_neurons
from sparse_autoencoder.train.sweep_config import SweepParametersRuntime


@final
class DefaultPipeline(AbstractPipeline):
    """Pipeline for training a Sparse Autoencoder on TransformerLens activations.

    Includes all the key functionality to train a sparse autoencoder, with a specific set of
        hyperparameters.
    """

    total_training_steps: int = 1

    def generate_activations(self, store_size: int) -> TensorActivationStore:
        """Generate activations."""
        num_neurons: int = 256
        model_device: torch.device = torch.device("mps")

        store = TensorActivationStore(store_size, num_neurons)

        # Set model to evaluation (inference) mode
        self.source_model.eval()
        self.source_model.to(model_device, print_details=False)

        # Add the hook to the model (will automatically store the activations every time the model runs)
        self.source_model.remove_all_hook_fns()
        hook = partial(store_activations_hook, store=store)
        self.source_model.add_hook(self.cache_name, hook)

        # Loop through the dataloader until the store reaches the desired size
        with torch.no_grad():
            for batch in self.source_data:
                input_ids: BatchTokenizedPrompts = batch["input_ids"].to(model_device)
                self.source_model.forward(input_ids, stop_at_layer=self.layer + 1)  # type: ignore (TLens is typed incorrectly)

                if len(store) >= store_size:
                    break

        self.source_model.remove_all_hook_fns()
        store.shuffle()

        return store

    def train_autoencoder(
        self, activation_store: TensorActivationStore, train_batch_size: int
    ) -> NeuronActivity:
        """Train the sparse autoencoder."""
        # TODO: move these to config
        model_device: torch.device = torch.device("mps")

        activations_dataloader = DataLoader(
            activation_store,
            batch_size=train_batch_size,
        )

        learned_activations_fired_count: NeuronActivity = torch.zeros(
            self.autoencoder.n_learned_features, dtype=torch.int32, device=model_device
        )

        for store_batch in activations_dataloader:
            # Zero the gradients
            self.optimizer.zero_grad()

            # Move the batch to the device (in place)
            batch = store_batch.detach().to(model_device)

            # Forward pass
            learned_activations, reconstructed_activations = self.autoencoder(batch)

            # Get metrics
            total_loss, loss_metrics = self.loss.batch_scalar_loss_with_log(
                batch, learned_activations, reconstructed_activations
            )

            # for metric in self.train_metrics:
            # TODO: Handle train metrics

            # Store count of how many neurons have fired
            with torch.no_grad():
                fired = learned_activations > 0
                learned_activations_fired_count.add_(fired.sum(dim=0))

            # Backwards pass
            total_loss.backward()
            self.optimizer.step()

            # Log
            # TODO: Make this commit at some point
            # if step % log_interval == 0 and wandb.run is not None:
            if wandb.run is not None:
                wandb.log(data=loss_metrics, step=self.total_training_steps)
            self.total_training_steps += 1

        return learned_activations_fired_count

    def validate_sae(self) -> None:
        """Get validation metrics."""
        # Not currently setup
        return

    def save_checkpoint(self) -> None:
        """Save the model as a checkpoint."""
        # Not currently setup
        return
