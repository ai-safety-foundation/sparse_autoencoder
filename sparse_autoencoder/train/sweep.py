"""Sweep."""
from pathlib import Path

import torch
from transformer_lens import HookedTransformer
from transformer_lens.utils import get_act_name, get_device
import wandb

from sparse_autoencoder import (
    ActivationResampler,
    AdamWithReset,
    L2ReconstructionLoss,
    LearnedActivationsL1Loss,
    LossReducer,
    Pipeline,
    PreTokenizedDataset,
    SparseAutoencoder,
)
from sparse_autoencoder.train.sweep_config import (
    RuntimeHyperparameters,
    SweepConfig,
)


def setup_activation_resampler(hyperparameters: RuntimeHyperparameters) -> ActivationResampler:
    """Setup the activation resampler for the autoencoder.

    Args:
        hyperparameters: The hyperparameters dictionary.

    Returns:
        ActivationResampler: The initialized activation resampler.
    """
    return ActivationResampler(
        resample_interval=hyperparameters["activation_resampler"]["resample_interval"],
        max_resamples=hyperparameters["activation_resampler"]["max_resamples"],
        n_steps_collate=hyperparameters["activation_resampler"]["n_steps_collate"],
        resample_dataset_size=hyperparameters["activation_resampler"]["resample_dataset_size"],
        dead_neuron_threshold=hyperparameters["activation_resampler"]["dead_neuron_threshold"],
    )


def setup_source_model(hyperparameters: RuntimeHyperparameters) -> HookedTransformer:
    """Setup the source model using HookedTransformer.

    Args:
        hyperparameters: The hyperparameters dictionary.

    Returns:
        The initialized source model.
    """
    return HookedTransformer.from_pretrained(
        hyperparameters["source_model"]["name"],
        dtype=hyperparameters["source_model"]["dtype"],
    )


def setup_autoencoder(
    hyperparameters: RuntimeHyperparameters, device: torch.device
) -> SparseAutoencoder:
    """Setup the sparse autoencoder.

    Args:
        hyperparameters: The hyperparameters dictionary.
        device: The computation device.

    Returns:
        The initialized sparse autoencoder.
    """
    autoencoder_input_dim: int = hyperparameters["source_model"]["hook_dimension"]
    expansion_factor = hyperparameters["autoencoder"]["expansion_factor"]
    return SparseAutoencoder(
        n_input_features=autoencoder_input_dim,
        n_learned_features=autoencoder_input_dim * expansion_factor,
    ).to(device)


def setup_loss_function(hyperparameters: RuntimeHyperparameters) -> LossReducer:
    """Setup the loss function for the autoencoder.

    Args:
        hyperparameters: The hyperparameters dictionary.

    Returns:
        The combined loss function.
    """
    return LossReducer(
        LearnedActivationsL1Loss(
            l1_coefficient=hyperparameters["loss"]["l1_coefficient"],
        ),
        L2ReconstructionLoss(),
    )


def setup_optimizer(
    autoencoder: SparseAutoencoder, hyperparameters: RuntimeHyperparameters
) -> AdamWithReset:
    """Setup the optimizer for the autoencoder.

    Args:
        autoencoder: The sparse autoencoder model.
        hyperparameters: The hyperparameters dictionary.

    Returns:
        The initialized optimizer.
    """
    return AdamWithReset(
        params=autoencoder.parameters(),
        named_parameters=autoencoder.named_parameters(),
        lr=hyperparameters["optimizer"]["lr"],
        betas=(
            hyperparameters["optimizer"]["adam_beta_1"],
            hyperparameters["optimizer"]["adam_beta_2"],
        ),
        weight_decay=hyperparameters["optimizer"]["adam_weight_decay"],
        amsgrad=hyperparameters["optimizer"]["amsgrad"],
        fused=hyperparameters["optimizer"]["fused"],
    )


def setup_source_data(hyperparameters: RuntimeHyperparameters) -> PreTokenizedDataset:
    """Setup the source data for training.

    Args:
        hyperparameters: The hyperparameters dictionary.

    Returns:
        PreTokenizedDataset: The initialized source data.
    """
    return PreTokenizedDataset(
        dataset_path=hyperparameters["source_data"]["dataset_path"],
        context_size=hyperparameters["source_data"]["context_size"],
    )


def setup_wandb() -> RuntimeHyperparameters:
    """Initialise wandb for experiment tracking."""
    wandb.init(project="sparse-autoencoder")
    return dict(wandb.config)  # type: ignore


def run_training_pipeline(
    hyperparameters: RuntimeHyperparameters,
    source_model: HookedTransformer,
    autoencoder: SparseAutoencoder,
    loss: LossReducer,
    optimizer: AdamWithReset,
    activation_resampler: ActivationResampler,
    source_data: PreTokenizedDataset,
) -> None:
    """Run the training pipeline for the sparse autoencoder.

    Args:
        hyperparameters: The hyperparameters dictionary.
        source_model: The source model.
        autoencoder: The sparse autoencoder.
        loss: The loss function.
        optimizer: The optimizer.
        activation_resampler: The activation resampler.
        source_data: The source data.
    """
    checkpoint_path = Path("../../.checkpoints")
    checkpoint_path.mkdir(exist_ok=True)

    random_seed = hyperparameters["random_seed"]
    torch.random.manual_seed(random_seed)

    hook_point = get_act_name(
        hyperparameters["source_model"]["hook_site"], hyperparameters["source_model"]["hook_layer"]
    )
    pipeline = Pipeline(
        activation_resampler=activation_resampler,
        autoencoder=autoencoder,
        cache_name=hook_point,
        checkpoint_directory=checkpoint_path,
        layer=hyperparameters["source_model"]["hook_layer"],
        loss=loss,
        optimizer=optimizer,
        source_data_batch_size=hyperparameters["pipeline"]["source_data_batch_size"],
        source_dataset=source_data,
        source_model=source_model,
        log_frequency=hyperparameters["pipeline"]["log_frequency"],
    )

    pipeline.run_pipeline(
        train_batch_size=hyperparameters["pipeline"]["train_batch_size"],
        max_store_size=hyperparameters["pipeline"]["max_store_size"],
        max_activations=hyperparameters["pipeline"]["max_activations"],
        checkpoint_frequency=hyperparameters["pipeline"]["checkpoint_frequency"],
        validate_frequency=hyperparameters["pipeline"]["validation_frequency"],
        validation_number_activations=hyperparameters["pipeline"]["validation_number_activations"],
    )


def sweep(sweep_config: SweepConfig) -> None:
    """Main function to run the training pipeline with wandb hyperparameter sweep."""
    sweep_id = wandb.sweep(sweep_config.to_dict(), project="sparse-autoencoder")

    def train() -> None:
        """Train the sparse autoencoder using the hyperparameters from the WandB sweep."""
        # Set up WandB
        hyperparameters = setup_wandb()

        # Setup the device for training
        device = get_device()

        # Set up the source model
        source_model = setup_source_model(hyperparameters)

        # Set up the autoencoder
        autoencoder = setup_autoencoder(hyperparameters, device)

        # Set up the loss function
        loss_function = setup_loss_function(hyperparameters)

        # Set up the optimizer
        optimizer = setup_optimizer(autoencoder, hyperparameters)

        # Set up the activation resampler
        activation_resampler = setup_activation_resampler(hyperparameters)

        # Set up the source data
        source_data = setup_source_data(hyperparameters)

        # Run the training pipeline
        run_training_pipeline(
            hyperparameters=hyperparameters,
            source_model=source_model,
            autoencoder=autoencoder,
            loss=loss_function,
            optimizer=optimizer,
            activation_resampler=activation_resampler,
            source_data=source_data,
        )

    wandb.agent(sweep_id, train)
