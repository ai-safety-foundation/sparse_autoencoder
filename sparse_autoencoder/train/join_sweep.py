"""Join an existing Weights and Biases sweep, as a new agent."""
import argparse

from sparse_autoencoder.train.sweep import sweep


def parse_arguments() -> argparse.Namespace:
    """Parse command line arguments.

    Returns:
        argparse.Namespace: Parsed command line arguments.
    """
    parser = argparse.ArgumentParser(description="Join an existing W&B sweep.")
    parser.add_argument(
        "--id", type=str, default=None, help="Sweep ID for the existing sweep.", required=True
    )
    return parser.parse_args()


def run() -> None:
    """Run the join_sweep script."""
    args = parse_arguments()

    sweep(sweep_id=args.id)


if __name__ == "__main__":
    run()
