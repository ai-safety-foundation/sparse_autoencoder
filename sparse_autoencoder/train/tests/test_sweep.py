"""Tests for sweep functionality."""

import pytest
from syrupy.session import SnapshotSession
import torch

from sparse_autoencoder.train.sweep import (
    setup_activation_resampler,
    setup_autoencoder,
    setup_loss_function,
)
from sparse_autoencoder.train.sweep_config import (
    RuntimeHyperparameters,
)


@pytest.fixture()
def dummy_hyperparameters() -> RuntimeHyperparameters:
    """Sweep config dummy fixture."""
    return {
        "activation_resampler": {
            "dead_neuron_threshold": 0.0,
            "max_resamples": 4,
            "n_steps_collate": 100_000_000,
            "resample_dataset_size": 819_200,
            "resample_interval": 200_000_000,
        },
        "autoencoder": {"expansion_factor": 4},
        "loss": {"l1_coefficient": 0.0001},
        "optimizer": {
            "adam_beta_1": 0.9,
            "adam_beta_2": 0.99,
            "adam_weight_decay": 0,
            "amsgrad": False,
            "fused": False,
            "lr": 1e-05,
        },
        "pipeline": {
            "checkpoint_frequency": 100000000,
            "log_frequency": 100,
            "max_activations": 2000000000,
            "max_store_size": 3145728,
            "source_data_batch_size": 12,
            "train_batch_size": 4096,
            "validation_frequency": 314572800,
            "validation_number_activations": 1024,
        },
        "random_seed": 49,
        "source_data": {"context_size": 128, "dataset_path": "NeelNanda/c4-code-tokenized-2b"},
        "source_model": {
            "dtype": "float32",
            "hook_dimension": 512,
            "hook_layer": 0,
            "hook_site": "mlp_out",
            "name": "gelu-2l",
        },
    }


def test_setup_activation_resampler(
    dummy_hyperparameters: RuntimeHyperparameters, snapshot: SnapshotSession
) -> None:
    """Test the setup_activation_resampler function."""
    activation_resampler = setup_activation_resampler(dummy_hyperparameters)
    assert snapshot == str(
        activation_resampler
    ), "Activation resampler string representation has changed."


def test_setup_autoencoder(
    dummy_hyperparameters: RuntimeHyperparameters, snapshot: SnapshotSession
) -> None:
    """Test the setup_autoencoder function."""
    autoencoder = setup_autoencoder(dummy_hyperparameters, device=torch.device("cpu"))
    assert snapshot == str(autoencoder), "Autoencoder string representation has changed."


def test_setup_loss_function(
    dummy_hyperparameters: RuntimeHyperparameters, snapshot: SnapshotSession
) -> None:
    """Test the setup_loss_function function."""
    loss_function = setup_loss_function(dummy_hyperparameters)
    assert snapshot == str(loss_function), "Loss function string representation has changed."
