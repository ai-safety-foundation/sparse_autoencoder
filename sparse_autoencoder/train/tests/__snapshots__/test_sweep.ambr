# serializer version: 1
# name: test_setup_activation_resampler
  'ActivationResampler(resample_interval=200000000, max_resamples=4, n_steps_collate=100000000, resample_dataset_size=819200, dead_neuron_threshold=0.0)'
# ---
# name: test_setup_autoencoder
  '''
  SparseAutoencoder(
    (_pre_encoder_bias): TiedBias(position=pre_encoder)
    (_encoder): LinearEncoder(
      in_features=512, out_features=2048
      (activation_function): ReLU()
    )
    (_decoder): UnitNormDecoder(in_features=2048, out_features=512)
    (_post_decoder_bias): TiedBias(position=post_decoder)
  )
  '''
# ---
# name: test_setup_loss_function
  '''
  LossReducer(
    (0): LearnedActivationsL1Loss(l1_coefficient=0.0001)
    (1): L2ReconstructionLoss()
  )
  '''
# ---
